### DISCUSSION TITLE
UICollector considers valid BG values invalid when CGM value is repeated

### DISCUSSION THREAD

**Discussion by sulkaharo** [NightscoutFoundation/xDrip#3108] [Bug reports]
I'm using the UiBasedCollector due to having switched to CamAPS and noticed the default behaviour is to assume if the notification sends the same value consecutively, the BG value is not valid, so it's not collected. This is causing us to miss CGM data frequently, as we have sufficiently stable BGs to have up to 8 or so repeatedly same glucose values. I'm assuming there must have been behaviour where the notifications repeat values, causing this logic to be added - or is this speculative? At least for my use, I'd like to do something like PR a config that allows me to have xDrip collect repeat values. Related line here: 

https://github.com/NightscoutFoundation/xDrip/blob/39370edfa35b5ce79fe5fba2c17dd1cffaed09db/app/src/main/java/com/eveningoutpost/dexdrip/services/UiBasedCollector.java#L304

**Comment by Navid200** [comment]
Thanks for opening this.

We will have to wait for JamOrHam.

I would like to say a few words.  I can think of a dangerous situation:
If an app keeps showing the same notification not because the glucose is stable, but, because the sensor has failed, xDrip has no way of knowing that.  If xDrip keeps showing the same reading, you may think that your glucose is still high after an injection when in reality, after 35 minutes (8 consecutive readings with 5 minute gaps in between), your glucose may actually be very low and you may need treatment.

This is a very dangerous situation we have to worry about.
So, when I look at this existing exception, I cannot help worry:  https://github.com/NightscoutFoundation/xDrip/blob/39370edfa35b5ce79fe5fba2c17dd1cffaed09db/app/src/main/java/com/eveningoutpost/dexdrip/services/UiBasedCollector.java#L344  

So, we can add another exception for CamAPS.  But, I hope we can come up with a safe approach.  

In my opinion, the companion app mode should not be advertised as an option.  it should be advertised as a last-resort until you figure out how to use xDrip as a collector or safe follower.

Let's see what JamOrHam thinks.

**Comment by sulkaharo** [comment]
Right, I wonder if the feature is there for the Libre. FWIW I've never seen this stuck sensor behaviour with Dexcom sensors CamAPS uses.

**Comment by Navid200** [comment]
@jamorham

**Comment by Navid200** [comment]
@sulkaharo Would you please try the latest Nightly (https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.10.24) to see if there is any improvement?

**Comment by botanicz** [comment]
Hello navid , i tried your latest nightly but unfortunately  it did not solve the problem described by sulkaharo...... 
we use cam aps and spread the bg values via xdrip in companion mode to the rest of the family with IOS devices . The problem for us is exactly like sulkaharo mentioned . Whenever ,my daughters BG vlaues are constant we have missing  values . When the values differ it collects every single value without any problem , but if the value is repeated it misses until the values differ again . I would be very grateful if you could help me/us with this topic !

**Reply by Navid200** [reply]
@botanicz What xDrip release are you using now?

**Reply by botanicz** [reply]
https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.10.24

**Reply by Navid200** [reply]
Do you mind showing a screenshot of the main screen where the readings have stopped?

The reason I ask is that I know there has been a change.  It used to be we would stop showing results if there were more than 4 consecutive identical readings.
After the change in the latest Nightly (October 24), we will stop showing results if there have been more than 6 consecutive identical readings.

If you show a screenshot, it may help us see what's going on exactly.

**Reply by botanicz** [reply]
![Screenshot_20231102_163245_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/83694871/d476dd0d-fe3c-4563-b9f1-b0f604c867dc)

**Reply by botanicz** [reply]
![Screenshot_20231102_164104_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/83694871/7954deca-0bdc-4728-943a-b27185bfab07)

**Reply by Navid200** [reply]
But, looking at that image, you can see the problem is not what this thread is about.
Have you enabled Bluetooth?
Please see the last heading on this page: https://navid200.github.io/xDrip/docs/Follow/CompanionApp.html

**Reply by botanicz** [reply]
yes ! enabled Bluetooth , when you look at the second screenshot you can clearly see that her BG valuue between 03.00 and 05.00 were mostly 42 the values were skipped ! all the values were displayed in camaps correctly but were not transferred to xdrip except the first one

**Reply by botanicz** [reply]
in my case back filling the values is not what is solving the problem , the real problem is xdrip skipping the same BG value coming from camaps ...... so the question is why is xdrip skipping notifications from camaps while camaps has all the values without skipping anything ..... ?

**Reply by botanicz** [reply]
![Screenshot_20231102_165214_CamAPS FX](https://github.com/NightscoutFoundation/xDrip/assets/83694871/306f27fc-932f-4e92-abe7-05dbb4dde8c1)

**Reply by botanicz** [reply]
above screenshot is showing the same time as the last xdrip screenshot ..... i was awake that night and my girlfriend did not get the values while me using camaps companion app got all the values .....

**Reply by Navid200** [reply]
Enable data tables on the master xDrip: https://navid200.github.io/xDrip/docs/Datatables.html
Then, scroll all the way back to get to where the missing readings are.  What does the data table show?

**Reply by botanicz** [reply]
hey navid ! i looked into the data table on the master xdrip ...... here a example screenshots i made at a different point but it always the same pattern when readings are not shown .... 
![IMG-20231103-WA0002](https://github.com/NightscoutFoundation/xDrip/assets/83694871/463d5a4a-5740-46b3-b34d-87337fb17ebb)

![IMG-20231103-WA0001](https://github.com/NightscoutFoundation/xDrip/assets/83694871/b39ab445-12b3-47f4-8e66-003067affc27)

**Reply by botanicz** [reply]
You see three times 151 around 9 o'clock . the value 151 right at 09:00 is missing in the graph but is shown in the data table ..... this is similar at all gaps in the graph (whenever the value is similar to the one before ) .... or do i understand it wrong ?

**Reply by Navid200** [reply]
@botanicz I have already explained why xDrip sops showing results if they don't change.  This is a safety measure because we have no way of knowing if the readings are real or if the app is malfunctioning and is just repeating the last reading it received.

This measure will trigger, as of the last xDrip release, only if there are 6 consecutive identical readings.  Previously, it would trigger after there were 4 consecutive identical readings.

I'm not sure what your data table shows G5 Native as the source of data.
What is the hardware data source set to?

**Reply by botanicz** [reply]
I understand so far but why did it trigger yesterday at 09:00 when there werde just three identical readings .....? I'm curious now or i understand it completly wrong ... sorry for any inconvinience i cause . I'm not sure where i cann see the hardware data source but in the status page the source is : UI BASED and in my settings the source is companion app .

**Reply by botanicz** [reply]
Would it be possible to provide a function to turn off this safety measure on user's risk ? For cases in which the measure is causing missing readings like we have ?

**Reply by Navid200** [reply]
It is possible to add more and more and more settings to xDrip.

It's not an easy task to fly a fighter jet.  But, that's not a problem because not everyone needs to fly a fighter jet.
But, everyone can get diabetes.  Everyone's child can get diabetes.   I don't want xDrip to become more and more complicated to use.  
The exact opposite.  I would like to clean up the settings in xDrip.
I want it to be simple so that anyone can use it without having to solve a differential equation first!

Please don't ask for new settings.
Settings should be added to xDrip if/when absolutely necessary.

What is your sensor?  Is it a Dexcom G6?

**Reply by botanicz** [reply]
It's a dexcom g6 that's right !

**Reply by Navid200** [reply]
And you must use CamAPS because of a pump?

**Comment by botanicz** [comment]
Yes ! Ypsopump ....

Navid ***@***.***> schrieb am Fr., 3. Nov. 2023, 20:59:

**Reply by Navid200** [reply]
For some Reason CamAPS decided to occupy the mobile slot of the G6 transmitter instead of its medical slot considering it's purpose is to facilitate a pump.   The t:slim pump occupies the medical slot and xDrip or Dexcom app can run in parallel.

By doing this, CamAPS have blocked any mobile app, even the official Dexcom app, from being operational.  
The culprit here, is CamAPS, not xDrip.
We accommodate and keep adding features to help.  

Looking at your readings, it doesn't look like the missing points are caused by the jam avoidance mechanism since your readings don't show 4 consecutive identical results when you have the missed reading.  So, even if we add the setting you are asking for, it's not going to make  difference at least for the case you are showing.

Let's wait for @jamorham to see if he has any idea what the cause of the missed readings could be.

**Comment by jamorham** [comment]
Are you using the Bluetooth companion mode enabled as well?

If the bluetooth mode doesn't fix it then I would say we probably will need to enable verbose logging for the collector and catch a moment in time where the problem exists and see if there are any clues in there.

**Reply by Navid200** [reply]
What I don't understand is why xDrip main screen is not showing something that is in the data table.
I didn't think that was possible.

**Reply by botanicz** [reply]
Hello @jamorham  ! at the moment bluetooth companion is enabled ..... a while ago i tried it without but there was no change in behaviour no matter bt companion mode is on or off . 
*edit : i need a quick hint where to find this verbose logging .if i should turn it on ..... thank you

**Comment by tzachi-dar** [comment]
@botanicz Do you have a valid example where the BG exists but is not on the screen (with 3 points)?
If so, can you export your db and send it to tzachi.dar@gmail.com and also state the time that the problem was happening?
